### ISSUE TITLE
Support React 16.3

### ISSUE THREAD

**Issue by theKashey** [gaearon/react-hot-loader#918]
Currently, we rely on `componentWillReceiveProps` to perform sync update and sync `forceUpdate`. So - this method does not exist anymore.

We have 2 options:
 1 - rely on subrender, ie async updates. It is a dev mode, nobody cares about double rendering on rare hot updates. Thus remove AppContainer completely. This will also solve 
 2 - find another way. But I don't see any.

**Comment by oliviertassinari** [comment]
Would it also explain why I see a warning anytime is use `getDerivedStateFromProps()`? I'm assuming react-hot-loader is hijacking the components and add `componentWillReceiveProps `.
![capture d ecran 2018-03-31 a 13 20 33](https://user-images.githubusercontent.com/3165635/38162706-8606e736-34e6-11e8-8b7b-3b48bbccaba8.png)

**Comment by theKashey** [comment]
++ we have to test how new ("real") Context API works, as long "Context" is not a "Tag" or a function - it is a brand new element type. Should be fine, but even enzyme itself failing :)

**Comment by theKashey** [comment]
@oliviertassinari - yeah, RHL is adding componentWillReceiveProps, and I am going just to drop it.

**Comment by swashata** [comment]
Had just this issue right now. I am very new to react, just started and had to scratch my head (like a lot) to understand this isn't coming from mine, rather the hot loader 😁 . I will be waiting for an update.

**Comment by gaearon** [comment]
I don't understand what this means:

>rely on subrender, ie async updates.

**Comment by theKashey** [comment]
@gaearon - this is a local meme( #830 ), about performing _hot-render_ each time you "can do it". V3 uses AppContainer to start an update, v4 can use AppContainer to perform sync updates, and Component.render to perform async updates. Async cos they could be executed after the main update or main update could not "render into them" - code splitting, resource load, portals, and so on. Plus we have to defer .forceUpdate, as long shall not call it from a render method.
Basically - this one line - https://github.com/gaearon/react-hot-loader/blob/master/src/proxy/createClassProxy.js#L164

**Comment by gaearon** [comment]
What are you using `componentWillReceiveProps` for today?

**Comment by theKashey** [comment]
Nothing but to being able "flush"(forceUpdate) changes. As I pointed above - we could safely remove it. It may cause double renders, but this is not something we should avoid.

**Comment by gaearon** [comment]
Got it. `componentWillReceiveProps` is a bad place to call `forceUpdate` so that sounds good.

**Comment by theKashey** [comment]
PS: Is there any "new"(~fiber) way to trigger re-render? Is there any way to re-render SFC without a real Component on top of it? Is there any a bit more "soft" ways to update?

**Comment by gaearon** [comment]
Can you be more specific? Maybe a small code example?

**Comment by theKashey** [comment]
- We found component to be updated.
 - We updated it.
 - Now we have to redraw it.

Do we have other choice rather then calling .forceUpdate? How do new Context API handle it?
Actually, what we are trying to do every time - is to update _everything_ from some point and beyond. Just _everything_ will also be a good solution.

**Comment by gaearon** [comment]
Isn't that what `react-force-deep-update` already doing? Do you need something else?

**Comment by theKashey** [comment]
Yep, this is related to `react-force-deep-update`. I don't really need anything else, but have to ask you about possibilities.

**Comment by gaearon** [comment]
React maintains a tree of nodes per root. Each node in a tree has an "expiration time" which can either be some time in the future, a marker that says "nothing to do here" or a marker that says "flush it synchronously".

When we `forceUpdate` or `setState`, what really happens is:

1. We also push its new state value into node's "update queue" (or set a flag for forced update)
2. We mark the related node as having "expiration time" of "synchronous"
3. We mark all its parent nodes up to the root the same way
4. We schedule React to "work" on this root

React loops over all scheduled roots. It contains this root. It then looks at all nodes that have expiration time that isn't "nothing to do here". It goes over each ancestor. It "skips over" the deep parents (before `setState` / `forceUpdate`) because they don't have anything new in the update queue and therefore they're safe to *not* re-render. Eventually it gets to the child that actually has something new in the update queue (our node) and that's when it starts re-rendering.

I hope this gives some insight into how it works internally. We could expose some of this, or we could keep using hacky internal APIs, or a mix of both. The question is what would be valuable to expose. I don't remember enough about how RHL works to say what would be, but maybe you could tell me?

Does this help at all?

**Comment by theKashey** [comment]
Look like we still have to use forceUpdate to mark node to be updated. We could, but shouldn't use setState for this, as long it may alter the way Component work.
And this is a less hacky way for now.

> I don't remember enough about how RHL works to say what would be, but maybe you could tell me?
- On HMR we call .forceUpdate on AppContainer bound to updated branch.
- Everything starts on update, in the render method after this issue being solved.
- RHL checks that "this.version" is the same as "current" value(number of reactHotLoader.register calls), and if not - trigger _hot-replacement-render_.
 - hot-render hydrates the tree and next start rendering component, comparing to hydrated or _registered_ components, performing proxies update then possible, and calling componentWillUpdate(to bypass redux component cache)
 - everything ends with another (forced) update, to actually display all the changes to the customer, bypassing PureComponents.

Ie - double, but in real triple, render. As long extra time, it will consume, is not comparable to the rebuild time - this is fine.

PS: And thank you for a conversation, I've just understood that comparing this.version and "current" is the root cause of #923, as long current does not increment without babel plugin.

**Comment by jaydenseric** [comment]
@gaearon can this please be elevated from an "enhancement" to a "bug"? Current stable React is meant to be supported according to [peer deps](https://github.com/gaearon/react-hot-loader/blob/v4.0.1/package.json#L93), yet here is the sort of console output we have to ignore:

<img width="605" alt="screen shot 2018-04-13 at 12 57 34 pm" src="https://user-images.githubusercontent.com/1754873/38714882-a44c5fd4-3f1b-11e8-8b9c-7cd5e9f70697.png">

These false positives are a big confusion and obfuscate real warnings that are really helpful in this period of learning and transition to the new component API. This issue will affect the DX for a lot of people; `react-hot-loader` has nearly half a million downloads/wk and is a dependency of Next.js.

Is there something blocking https://github.com/gaearon/react-hot-loader/pull/927?

**Comment by theKashey** [comment]
Released. v4.1.0

**Comment by mvestergaard** [comment]
4.1.0 got rid of warnings related to `componentWillReceiveProps`, which is great.  
However there still seems to be the same issue related to `componentWillUpdate`.

Here's a warning i receive while using https://github.com/reactjs/react-modal. It does not as far as I can tell implement `componentWillUpdate`, so I'm assuming it's coming from the proxy.
```
Warning: Unsafe legacy lifecycles will not be called for components using new component APIs.

Modal uses getSnapshotBeforeUpdate() but also contains the following legacy lifecycles:
  componentWillUpdate

The above lifecycles should be removed. Learn more about this warning here:
https://fb.me/react-async-component-lifecycle-hooks
```

**Comment by theKashey** [comment]
@mvestergaard  - double checked - there is no `componentWillReceiveProps` in `src` folder. Could you double check?

**Comment by mvestergaard** [comment]
@theKashey I think you misunderstood. It's not related to `componentWillReceiveProps`.  
The `Modal` component uses `getSnapshotBeforeUpdate`, see: https://github.com/reactjs/react-modal/blob/master/src/components/Modal.js#L123

But does not use `componentWillUpdate`, however the above warning appears.

**Comment by theKashey** [comment]
I’ve just copy-pasted wrong function. 
Anyway - I’ve searched React-hot-loader sources and there is nothing related to deprecated methods. 
That might be due ton polyfill, as long it adds all the deprecated methods, ignoring current React version, but setting secret flag to suppress the warnings. 
And next we wrapping all the methods, but not passing that secret flags. 
Thus a it is know how to fix it.

**Comment by gaearon** [comment]
@mvestergaard If you create a small reproducing example I can look.

**Comment by gaearon** [comment]
I think the issue is that React uses a flag on the method to determine whether it's coming from `react-lifecycles-compat` polyfill (in which case the method shouldn't lead to a warning). But proxying doesn't preserve the flag.

If I change [this method](https://github.com/gaearon/react-hot-loader/blob/948138ace84d08c95f29870b1e461154f1208cc7/src/proxy/createClassProxy.js#L118-L120) to include

```js
    Object.assign(wrappedMethod, realMethod);
```

before returning `wrappedMethod` then the issue goes away.

**Comment by theKashey** [comment]
I'll update proxy code to keep everything from original class. @gaearon - should I also keep method arity, or (I hope) argument count does not count?

**Comment by gaearon** [comment]
Arity shouldn’t matter although preserving it is always nice if it’s not too difficult.

**Comment by theKashey** [comment]
Patch version released. v4.1.1

**Comment by elisherer** [comment]
I'm working with React 16.3 and the HMR is broken in 4.1.0 and 4.1.1.
Changing even a simple string inside a div won't appear in the browser after module updates (in my case a function component, didn't check a class one).
rolling back to RHL 4.0.1 brings back the HMR.

**Comment by gaearon** [comment]
Can you provide a repro case?

**Comment by elisherer** [comment]
You caught me unprepared, it might take me a lot of time. My project is not small, but it doesn't use anything special either.
Some code I use:
### package.json
```json
  "scripts": {
    "start": "webpack-serve",
```

### .babelrc
```json
{
  "presets": [
    "@babel/preset-flow",
    "@babel/preset-react",
    ["@babel/preset-env", { "modules": false }]
  ],
  "plugins": [
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-object-rest-spread",
    "@babel/plugin-syntax-dynamic-import"
  ],
  "env": {
    "test": {
      "plugins": [
        "@babel/plugin-transform-modules-commonjs"
      ]
    },
    "development": {
      "plugins": [
        "react-hot-loader/babel"
      ]
    },
    "production": {
      "plugins": [
        "@babel/plugin-transform-react-constant-elements",
        "@babel/plugin-transform-react-inline-elements",
        "babel-plugin-transform-react-pure-class-to-function",
        "babel-plugin-transform-react-remove-prop-types"
      ]
    }
  }
}
```

### App.js
```js
...

let exportedApp = App;

if (__DEV__) {
  const { hot }  = require('react-hot-loader');
  exportedApp = hot(module)(App);
}

export default connect(mapStateToProps)(exportedApp);
```

### index.js
```js
import React from 'react';
import { render } from 'react-dom';
import App from './components/App';

import { Provider } from './store'; // store made with the new Context API

const appElement = document.getElementById('root');

render(
  <Provider>
    <App />
  </Provider>
, appElement);
```

**Comment by theKashey** [comment]
@elisherer - you can activate debug mode - https://github.com/gaearon/react-hot-loader#setconfigconfig - and it shall explain what is wrong.

**Comment by theKashey** [comment]
I've double checked and added a simple test to be confident about - everything is still 👌

**Comment by elisherer** [comment]
@theKashey, Sorry, I don't get any debug info besides the regular logs:
![image](https://user-images.githubusercontent.com/1045347/39060517-3ba53942-44ca-11e8-9de2-bfb49a312916.png)

**Comment by theKashey** [comment]
@elisherer - just to double check - App is included into `templates/index` or `templates/index` is included in App?

The main difference between 4.0.1 and 4.1 - SFCs are no longer converted into Stateful components, and as result - they lose "brains".
They could not found that "O! I was updated! Hot reloading", and have to had AppContainer(or `hot`), or any other "smart" Stateful component before them.

Your application worked because "**safety net**" worked. **I've removed it** from SFC, and it got broken.
I am not sure how you did it. And I could not reproduce it.

I'll be quite keen to get an example to play with. You can send it in private, if it is not public.

Also, you can try to restore that net - just add `App.contexTypes = {}`. SFCs with context are still have to be converted into Stateful ones with safety things enabled.

That safety net actually activates "React-Hot-Loader" if it did not properly do it's work. As long there are no errors - look like RHL is just doing nothing, ignoring the update.

**Comment by swashata** [comment]
Here's my two cents. I have followed the instruction and it is just working good with React 16.3.

**Foo.jsx**
```js
import React from 'react';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

class Foo extends React.PureComponent {
	static propTypes = {
		categories: PropTypes.arrayOf(PropTypes.number),
	};

	static defaultProps = {
		categories: [],
	};

	state = {
		foo: 'Bar',
	};
	render() {
		return (
			<div className="foo-container">{this.state.foo}</div>
		);
	}
}

export default hot(module)(Foo);
```

The only difference is that I haven't made an export hot, while connecting it with the store. I am using Redux though and my `App.jsx` looks different.

**App.jsx**
```js
import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';

import Foo from './Foo';
import store from './store';

const Root = (
	<Provider store={store}>
		<Foo />
	</Provider>
);

render(Root, document.getElementById('foo-app'));
```

And to have hot reloading within the store,

**store.js**
```js
import { createStore } from 'redux';
import rootReducers from './reducers/rootReducers';

// Create the store
const store = createStore(
	rootReducers,
	/* eslint-disable no-underscore-dangle */
	// Add devtool in an isomorphic way
	typeof window !== 'undefined' &&
		window.__REDUX_DEVTOOLS_EXTENSION__ &&
		window.__REDUX_DEVTOOLS_EXTENSION__()
	/* eslint-enable */
);

// webpack hot middleware support
if (module.hot) {
	module.hot.accept('./reducers/rootReducers', () => {
		/* eslint-disable global-require */
		const nextRootReducers = require('./reducers/rootReducers').default;
		/* eslint-enable */
		store.replaceReducer(nextRootReducers);
	});
}

// export the store
export default store;
```

And all these setup works with webpack dev middle and webpack hot middleware. Since I am developing a WP plugin, I couldn't use Webpack Dev server directly. So I used the middlewares inside browsersync (all connected with a simple gulp task).

**gulpfile.js**
```js
const gulp = require('gulp');
const browserSync = require('browser-sync').create();
const webpack = require('webpack');

const webpackDevMiddleware = require('webpack-dev-middleware');
const webpackHotMiddleware = require('webpack-hot-middleware');

// Get the function for creating varying config
const webpackConfig = require('./config/webpack.config');
// Our local config
const localConfig = require('./config/local.config');

gulp.task('serve', () => {
	const config = webpackConfig('development', localConfig);
	const webpackBundler = webpack(config);
	const devMiddleware = webpackDevMiddleware(webpackBundler, {
		noInfo: true,
		publicPath: config.output.publicPath,
		stats: { colors: true },
	});
	const devHotMiddleWare = webpackHotMiddleware(webpackBundler);
	browserSync.init({
		logLevel: 'info',
		port: localConfig.port,
		ui: localConfig.ui,
		debugInfo: true,
		proxy: {
			target: localConfig.proxy,
			// Middleware for webpack hot reload
			middleware: [
				// converts browsersync into a webpack-dev-server
				devMiddleware,
				// hot reload JS
				devHotMiddleWare,
			],
		},
		host: localConfig.host,
		open: false,
		notify: localConfig.notify,
		background: true,
	});

	// Full reload on php changes
	gulp.watch(PATHS.php.src).on('change', browserSync.reload);
	// JS & CSS will be handled by webpack on the go
});
```

The codes are stripped down so those are missing some imports.

I hope the concepts and codes above helps @elisherer to debug your issue. I don't think it has anything to do with react-hot-loader rather how you have implemented it.

Also, I guess you do have the `"react-hot-loader/babel"` plugin in your `.babelrc` file?

**Comment by swrobel** [comment]
@theKashey this is still open despite being listed as fixed in the 4.1.0 release notes. Is there a commit that fixed it?

**Comment by theKashey** [comment]
@swrobel it was "fixed" in 4.1.0, and yet again in "4.1.1" [4 days ago](https://github.com/gaearon/react-hot-loader/pull/939), and now I am waiting confirmation from @elisherer and @mvestergaard that _this time_ is everything ok.
> All the tests I've done myself - green.

**Comment by elisherer** [comment]
ill try to create a repro repo asap

**Comment by mvestergaard** [comment]
@theKashey my issue was resolved with 4.1.1, thank you.

**Comment by elisherer** [comment]
It has something to do with me using `react-waterfall` which isn't doing anything wrong (from my perspective).
Check it out on: https://github.com/elisherer/react-hot-loader-repro918
I added examples of:
* Class component
* Function component
* Pure Class component
* Consumer Class component
* Consumer Function component
* Consumer Pure Class component
* Children as a Function component
* Consumer "Connected" component
* Connected Children as a Function component

All is working well **besides** the `react-waterfall` example.
`react-waterfall` is a mixture of the above and still manages to break, I added the library's source into the repo so you could fiddle with it. 
I Couldn't get it to work.

**Comment by theKashey** [comment]
@elisherer - this is not due to this issue, and it is easy to "fix" it - just remove "Prevent" component.
That means - "forceUpdating" of PureComponents might be broken. That's another issue.

This one is ✅🤘
